### ISSUE TITLE
SHA256 fails for multiple 512-bit blocks

### ISSUE THREAD

**Issue by anijain2305** [secworks/sha256#6]
The testbench tb_sha256_core.v does not test the functionality of multiple blocks correctly. 

The issue in the core design is that the H*_reg and a-g reg values are set to reset everytime a new 512-bit block comes. The issue can be fixed by updating the if(next) portion of FSM in the core implementation. I am highlighting the changes here.

            if (init)
              begin
                digest_init      = 1;
                w_init           = 1;
                state_init       = 1;
                first_block      = 1;
                t_ctr_rst        = 1;
                digest_valid_new = 0;
                digest_valid_we  = 1;
                sha256_ctrl_new  = CTRL_ROUNDS;
                sha256_ctrl_we   = 1;
              end

            if (next)
              begin
                w_init           = 1;
                state_init       = 1;
                t_ctr_rst        = 1;
                **digest_init      = 0;
                first_block      = 0;**
                digest_valid_new = 0;
                digest_valid_we  = 1;
                sha256_ctrl_new  = CTRL_ROUNDS;
                sha256_ctrl_we   = 1;
              end

**Comment by secworks** [comment]
Hmmm, weird. Thanks for finding this and posting.

**Comment by anijain2305** [comment]
After reading my post again, I realize that I did not explain the issue properly yesterday. So, there are two issues here - one is in testbench and the testbench issue prevents us from finding a bug in the core.

1) tb_sha256_core.v - for double_block_test, the second check seems typo and it still compares the first block output (instead of second block). When I fixed the testbench, the output was incorrect.

2) sha256_core.v - the issue comes while resetting the values of hashes and the a-g reg values when the next block arrives. At that time, they are supposed to retain their values. Currently, it is getting reset to the values corresponding to the arrival of the first block. 

Let me know if you need more help with this.

**Comment by secworks** [comment]
Sounds like a nice double bug. I'll fix.

**Comment by secworks** [comment]
Fixed the core testbench, there were quite a few nits there. And indeed it did the wrong comparison.

I've also corrected the core so now it generates the correct digest for the multi block test. Note that I use default assignments in the FSM so I don't need to set digest_init and first block to zero in the next case since they have the correct value already.

The big problem was really the core-testbench that also didn't use the core correctly.

The fix is in: https://github.com/secworks/sha256/commit/8076da96dc2a175d333d5c45d3bc7e664451285a
